sunxi-boot0.c

```diff
@@ -10,218 +11,640 @@
+		param->para0);
+	fprintf(stream, "dram_para1\t   = 0x%08X,\n", param->para1);
+	fprintf(stream, "dram_para2\t   = 0x%08X,\n", param->para2);
+	fprintf(stream, "dram_mr0\t   = 0x%X,\n", param->mr0);
+	fprintf(stream, "dram_mr1\t   = 0x%X,\n", param->mr1);
+	fprintf(stream, "dram_mr2\t   = 0x%X,\n", param->mr2);
+	fprintf(stream, "dram_mr3\t   = 0x%X,\n", param->mr3);
+	fprintf(stream, "dram_mr4\t   = 0x%X,\n", param->mr4);
+	fprintf(stream, "dram_mr5\t   = 0x%X,\n", param->mr5);
+	fprintf(stream, "dram_mr6\t   = 0x%X,\n", param->mr6);
+	fprintf(stream, "dram_mr11\t   = 0x%X,\n", param->mr11);
+	fprintf(stream, "dram_mr12\t   = 0x%X,\n", param->mr12);
+	fprintf(stream, "dram_mr13\t   = 0x%X,\n", param->mr13);
+	fprintf(stream, "dram_mr14\t   = 0x%X,\n", param->mr14);
+	fprintf(stream, "dram_mr16\t   = 0x%X,\n", param->mr16);
+	fprintf(stream, "dram_mr17\t   = 0x%X,\n", param->mr17);
+	fprintf(stream, "dram_mr22\t   = 0x%X,\n", param->mr22);
+	fprintf(stream, "dram_tpr0\t   = 0x%08X,\n", param->tpr0);
+	fprintf(stream, "dram_tpr1\t   = 0x%X,\n", param->tpr1);
+	fprintf(stream, "dram_tpr2\t   = 0x%X,\n", param->tpr2);
+	fprintf(stream, "dram_tpr3\t   = 0x%X,\n", param->tpr3);
+	fprintf(stream, "dram_tpr6\t   = 0x%08X,\n", param->tpr6);
+	fprintf(stream, "dram_tpr10\t   = 0x%08X,\n", param->tpr10);
+	fprintf(stream, "dram_tpr11\t   = 0x%08X,\n", param->tpr11);
+	fprintf(stream, "dram_tpr12\t   = 0x%08X,\n", param->tpr12);
+	fprintf(stream, "dram_tpr13\t   = 0x%X,\n", param->tpr13);
+	fprintf(stream, "dram_tpr14\t   = 0x%X, ; should be 0 on anything but A523\n",
+		param->tpr14);
+	fprintf(stream, "\n");
+}
+
+static void
+dram_param_raw_print(FILE *stream, void *raw)
+{
+	uint32_t *param = raw;
+	int i;
+
+	fprintf(stream, "; Unknown structure\n");
+	for (i = 0; i < BOOT0_DRAM_PARAM_COUNT; i++)
+		fprintf(stream, "dram_%02d\t= 0x%08X\n", i, param[i]);
+}
+
+/*
+ * If we return 0, we tell the upper layer to skip only 1 sector.
+ */
+int output_boot0_info(void *sector, FILE *inf, FILE *stream, bool verbose)
+{
+	struct egon_header *header = sector;
+	struct egon_header_secondary *secondary;
+	void *dram_param;
+	int ret;
+
+	if (memcmp(header->magic, BOOT0_MAGIC_0, 8) > 0) {
+		fprintf(stream,
+			"\tERROR: wrong header magic: %c%c%c%c%c%c%c%c\n",
+			header->magic[0], header->magic[1],
+			header->magic[2], header->magic[3],
+			header->magic[4], header->magic[5],
+			header->magic[6], header->magic[7]);
+		return 0;
+	}
+
+	if (header->header_size != sizeof(struct egon_header)) {
+		fprintf(stream, "\tERROR: egon header size mismatch: %d\n",
+			header->header_size);
+		return 0;
+	}
+
+	if (header->filesize & (BOOT0_FILESIZE_ALIGN - 1)) {
+		fprintf(stream, "\tERROR: boot0 file size not a multiple of "
+			"%d: %d bytes (0x%04X).\n", BOOT0_FILESIZE_ALIGN,
+			header->filesize, header->filesize);
+		return 0;
+	}
+
+	/* Return to start of the boot0 data. */
+	ret = fseek(inf, -512, SEEK_CUR);
```

Comment:
No. The whole program is designed to work on a stream, not only on seekable files. So no backwards seeking. The input file could be the output of a decompressor, or piped in via network.
The idea is that the sector that was just read is stored in the buffer `*sector` points to. If you need more data, you can read more or seek forward, but otherwise just use what's already there.

Reply:
Ok. That explains why mmap() was not used, which is my preferred usage of anything binary.